pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,400 @@
+            case float():
+                self.lib.conduit_node_set_path_float32(self, key, value)
+            case (np.ndarray() | np.generic()):
+                value = np.ascontiguousarray(value)
+                fn = getattr(self.lib, 
+                             f'conduit_node_set_path_{value.dtype}_ptr')
+                fn(self, key, value.ctypes.data, value.size)
+            case _:
+                ValueError(f'ConduitNode: __setitem__ type not supported')
+
+    def __str__(self):
+        self.lib.conduit_node_print(self)
+        return '\n'
+
+    def append(self):
+        ptr = self.lib.conduit_node_append(self)
+        return ConduitNode(self.lib, ptr, child=True)
+
+
+class AscentWrappers(LibWrapper):
+    _libname = 'ascent_mpi'
+
+    # Functions
+    _functions = [
+        (None, 'ascent_close', c_void_p),
+        (c_void_p, 'ascent_create', c_void_p),
+        (None, 'ascent_execute', c_void_p, c_void_p),
+        (None, 'ascent_open', c_void_p, c_void_p),
+        (None, 'ascent_publish', c_void_p, c_void_p),
+    ]
+
+
+class AscentPlugin(RegionMixin, BasePlugin):
+    name = 'ascent'
+    systems = ['*']
+    formulations = ['dual', 'std']
+
+    # Ascent filters that implicit require velocity expression
+    v_filter = ['gradient', 'qcriterion', 'vorticity']
+
+    def __init__(self, intg, cfgsect, suffix=None):
+        super().__init__(intg, cfgsect, suffix)
+
+        comm, rank, root = get_comm_rank_root()
+        self.nsteps = self.cfg.getint(cfgsect, 'nsteps')
+
+        # Get datatype
+        dtype = intg.system.backend.fpdtype
+
+        # Underlying system elements class
+        self.elementscls = intg.system.elementscls
+
+        # Load conduit library
+        self.conduit = ConduitWrappers()
+
+        # Expressions to plot and configs
+        self._exprs = defaultdict(dict)
+        self._fields_write = set()
+        self._fields_read = set()
+        self._config_field(cfgsect)
+        self._config_scene(cfgsect)
+        self._config_pipeline(cfgsect)
+
+        if not self._fields_read.issubset(self._fields_write):
+            raise AscentError('Not all fields used are defined')
+        
+        # Gradient pre-processing
+        self._init_gradients()
+
+        # Base directory for output and counter reset
+        basedir = self.cfg.getpath(cfgsect, 'basedir', '.', abs=True)
+
+        for scfg in self._scene_cfg.values():
+            scfg['path_gen'] = file_path_gen(basedir, scfg['image_name'],
+                                             intg.isrestart)
+            next(scfg['path_gen'])
+
+        # Set subdivision
+        self.divisor = self.cfg.getint(cfgsect, 'division', 
+                                       self.cfg.getint('solver', 'order'))
+        self.etypes_div = defaultdict(lambda: self.divisor)
+
+        # Share number of ele types per part
+        self.dom_0 = dom_0 = comm.exscan(len(intg.system.ele_map)) or 0
```

Comment:
Can this return None?  Also check the length makes sense in the context of regions (where you may only have some element types inside the region).